### ISSUE TITLE
Trouble when multiple widget have the same name

### ISSUE THREAD

**Issue by sghpjuikit** [sghpjuikit/player#150]
Infinite recompilation cycle. Can happen when user clones a widget to use it as a starting point.

**Comment by sghpjuikit** [comment]
Also causes widgets opened through templates to not show gui compilation progress

**Comment by sghpjuikit** [comment]
Infinite recompilation cycle fixed in a2e1d87f by asserting compilation result (class name and package name must match widget directory name).

The problem with ids has been improved in e9d79001: Widget.name is no longer used for identifying purposes, instead Widget.factoryId has been introduced.

I'm still planning to turn factoryId into UUID, rather than text containing the widget directory name, to prevent collisions, but at this point I can finally sleep at peace.

**Comment by sghpjuikit** [comment]
I fixed related issue: unrecognized widgets created using extension `WidgetFactory.orNone()` did not reload after recompilation nor show recompilation.

The former was caused by using name instead of factoryId (see above), while the latter stemmed from the issue of recompilation identity - during recompilation time, factory may not exist (only factoryId) and thus it could not match any recompilation entry. Recompilation entry identity was changed to widget location. This ties nicely into `WidgetFactory.identityId`, which is tied to `Widget.controller` class name, which must conform to name of this location.

This completes the identity issues. Just to reiterate the various identities involved:
* widget name - defaults to widget factory's name, but can be overriden by user
* widget id - unique per widget instance id
* widget controller class name - derived from widget location
* widget factory name
* widget factory location - location of the widget
* widget factory id - unique per widget type id, derived from widget controller class name
* widget factory recompilation id - same as widget factory location

There are still minor improvements to be made, most important of which is the lack of empty content for components, in the same vein as for widgets, to allow to wait and reload when the component becomes available to the application.
